File: fastapi_book_server/app/serializers/sequence.py
from datetime import date
from typing import Optional

from fastapi_pagination import Page
from pydantic import BaseModel


class Sequence(BaseModel):
    id: int
    name: str


class Author(BaseModel):
    id: int

    first_name: str
    last_name: str
    middle_name: Optional[str]

    annotation_exists: bool


class Book(BaseModel):
    id: int
    title: str
    lang: str
    file_type: str
    available_types: list[str]
    uploaded: date
    authors: list[Author]
    translators: list[Author]
    annotation_exists: bool


class PageWithSequence(Page[Book]):
    parent_item: Sequence | None

File: src/serializers/sequence.rs
use chrono::NaiveDate;
use serde::Serialize;

use super::author::Author;
use super::date::naive_date_serializer;

#[derive(sqlx::FromRow, sqlx::Type, Serialize)]
#[sqlx(type_name = "sequence_type")]
pub struct Sequence {
    pub id: i32,
    pub name: String,
}

#[derive(sqlx::FromRow, Serialize)]
pub struct SequenceBook {
    pub id: i32,
    pub title: String,
    pub lang: String,
    pub file_type: String,
    pub year: i32,
    pub available_types: Vec<String>,
    #[serde(serialize_with = "naive_date_serializer::serialize")]
    pub uploaded: NaiveDate,
    pub authors: Vec<Author>,
    pub translators: Vec<Author>,
    pub annotation_exists: bool,
    pub position: i32,
}
